core/src/test/java/com/linecorp/armeria/client/proxy/HAProxyClientIntegrationTest.java

```diff
@@ -0,0 +1,393 @@
+import io.netty.handler.logging.LoggingHandler;
+import io.netty.util.ReferenceCountUtil;
+
+class HAProxyClientIntegrationTest {
+    private static final String PROXY_PATH = "/proxy";
+    private static final DynamicBehaviorHandler DYNAMIC_HANDLER = new DynamicBehaviorHandler();
+
+    @RegisterExtension
+    static ServerExtension backendServer = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.port(0, SessionProtocol.HTTP, SessionProtocol.PROXY);
+            sb.port(0, SessionProtocol.HTTPS, SessionProtocol.PROXY);
+            sb.tlsSelfSigned();
+            sb.service(PROXY_PATH, (ctx, req) -> {
+                assertThat(ctx.proxiedAddresses().destinationAddresses()).hasSize(1);
+                final String proxyString = String.format("%s-%s", ctx.proxiedAddresses().sourceAddress(),
+                                                         ctx.proxiedAddresses().destinationAddresses().get(0));
+                return HttpResponse.of(proxyString);
+            });
+        }
+    };
+
+    @RegisterExtension
+    static NettyServerExtension http1Server = new NettyServerExtension() {
+        @Override
+        protected void configure(Channel ch) throws Exception {
+            ch.pipeline().addLast(new LoggingHandler(getClass()));
+            ch.pipeline().addLast(new HAProxyMessageDecoder());
+            ch.pipeline().addLast(new HttpServerCodec());
+            ch.pipeline().addLast(new HttpObjectAggregator(1024));
+            ch.pipeline().addLast(DYNAMIC_HANDLER);
+        }
+    };
+
+    @ParameterizedTest
+    @ArgumentsSource(ProtocolEndpointProvider.class)
+    void testExplicitHAProxy(SessionProtocol sessionProtocol, InetSocketAddress proxyAddr) throws Exception {
+        final InetSocketAddress srcAddr = new InetSocketAddress("127.0.0.2", 82);
+        final InetSocketAddress destAddr = new InetSocketAddress("127.0.0.3", 83);
+        final Endpoint destEndpoint = Endpoint.of(destAddr.getHostString(), destAddr.getPort());
+
+        try (ClientFactory clientFactory =
+                     ClientFactory.builder()
+                                  .proxyConfig(ProxyConfig.haproxy(proxyAddr, srcAddr))
+                                  .tlsNoVerify()
+                                  .useHttp2Preface(true)
+                                  .build()) {
+
+            final WebClient webClient = WebClient.builder(sessionProtocol, destEndpoint)
+                                                 .factory(clientFactory)
+                                                 .decorator(LoggingClient.newDecorator())
+                                                 .build();
+            final CompletableFuture<AggregatedHttpResponse> responseFuture =
+                    webClient.get(PROXY_PATH).aggregate();
+
+            final AggregatedHttpResponse response = responseFuture.get(10, TimeUnit.SECONDS);
```

Comment:
nit: Could replace with join()`? Unless you intended to control a specific timeout, JUnit will do it for you.
https://github.com/line/armeria/blob/ad0350c15ea098d897657cc79df461ac1aded898/testing-internal/src/main/resources/junit-platform.properties#L1